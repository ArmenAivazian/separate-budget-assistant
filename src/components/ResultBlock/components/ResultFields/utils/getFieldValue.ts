export function getFieldValue(
  salaryArmen: string,
  salaryNastya: string,
  value: number
) {
  const percentDifferent = +salaryNastya / +salaryArmen;

  const firstValue = Math.round((1 - percentDifferent) * value);
  const secondValue = Math.round(percentDifferent * value);

  return firstValue ? { firstValue, secondValue } : null;
}
