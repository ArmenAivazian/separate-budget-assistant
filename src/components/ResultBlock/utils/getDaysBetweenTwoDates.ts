export function getDaysBetweenTwoDates(start: string, end: string) {
  return Math.ceil(
    Math.abs(+new Date(start) - +new Date(end)) / (1000 * 60 * 60 * 24)
  );
}
