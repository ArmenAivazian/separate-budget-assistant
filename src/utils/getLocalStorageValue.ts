export function getLocalStorageValue(name?: string) {
  if (!name) return "";

  const value = localStorage.getItem(name);

  if (value) return value;

  return "";
}
