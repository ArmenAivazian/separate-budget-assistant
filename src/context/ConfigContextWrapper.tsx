import { createContext, useState } from "react";
import { local } from "../local/en";
import { ConfigValue } from "../types/ConfigValue";
import { LocalStorageName } from "../types/LocalStorageName";
import { getLocalStorageValue } from "../utils/getLocalStorageValue";

export const ConfigContext = createContext<{
  armenSalary: ConfigValue;
  nastyaSalary: ConfigValue;
  rent: ConfigValue;
  investmentRatio: ConfigValue;
  firstDay: ConfigValue;
  costProducts: ConfigValue;
  lastDay: ConfigValue;
  costHealth: ConfigValue;
  costServices: ConfigValue;
  costCommunalApartment: ConfigValue;
  targetFirstDay: ConfigValue;
  targetLastDay: ConfigValue;
}>({} as any);

export function ConfigContextWrapper({ children }: { children: JSX.Element }) {
  const defArmenSalary = getLocalStorageValue(LocalStorageName.ArmenSalary);
  const defNastyaSalary = getLocalStorageValue(LocalStorageName.NastyaSalary);
  const defRent = getLocalStorageValue(LocalStorageName.Rent);
  const defInvestRatio = getLocalStorageValue(LocalStorageName.InvestmentRatio);

  const [armenSalary, setArmenSalary] = useState(defArmenSalary);
  const [nastyaSalary, setNastyaSalary] = useState(defNastyaSalary);
  const [rent, setRent] = useState(defRent);
  const [investmentRatio, setInvestmentRatio] = useState(defInvestRatio);
  const [firstDay, setFirstDay] = useState("");
  const [lastDay, setLastDay] = useState("");
  const [costProducts, setCostProducts] = useState("");
  const [costHealth, setCostHealth] = useState("");
  const [costServices, setCostServices] = useState("");
  const [costCommunalApartment, setCommunalApartment] = useState("");
  const [targetFirstDay, setTargetFirstDay] = useState("");
  const [targetLastDay, setTargetLastDay] = useState("");

  return (
    <ConfigContext.Provider
      value={{
        armenSalary: {
          text: local.salaryArmen,
          value: armenSalary,
          localStorageName: LocalStorageName.ArmenSalary,
          setValue: setArmenSalary,
        },
        nastyaSalary: {
          text: local.salaryNastya,
          value: nastyaSalary,
          localStorageName: LocalStorageName.NastyaSalary,
          setValue: setNastyaSalary,
        },
        rent: {
          text: local.rent,
          value: rent,
          localStorageName: LocalStorageName.Rent,
          setValue: setRent,
        },
        investmentRatio: {
          text: local.investmentRatio,
          value: investmentRatio,
          localStorageName: LocalStorageName.InvestmentRatio,
          setValue: setInvestmentRatio,
        },
        firstDay: {
          text: local.firstDayCalculate,
          value: firstDay,
          type: "date",
          setValue: setFirstDay,
        },
        lastDay: {
          value: lastDay,
          text: local.lastDayCalculate,
          type: "date",
          setValue: setLastDay,
        },
        costProducts: {
          text: local.products,
          value: costProducts,
          setValue: setCostProducts,
        },
        costHealth: {
          text: local.health,
          value: costHealth,
          setValue: setCostHealth,
        },
        costServices: {
          text: local.services,
          value: costServices,
          setValue: setCostServices,
        },
        costCommunalApartment: {
          text: local.communalApartment,
          value: costCommunalApartment,
          setValue: setCommunalApartment,
        },
        targetFirstDay: {
          text: local.targetFirstDay,
          value: targetFirstDay,
          type: "date",
          setValue: setTargetFirstDay,
        },
        targetLastDay: {
          text: local.targetLastDay,
          value: targetLastDay,
          type: "date",
          setValue: setTargetLastDay,
        },
      }}
    >
      {children}
    </ConfigContext.Provider>
  );
}
