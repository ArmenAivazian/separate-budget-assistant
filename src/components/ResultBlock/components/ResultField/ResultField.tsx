import { ConfigContext } from "../../../../context/ConfigContextWrapper";
import style from "./ResultField.module.css";
import { useContext } from "react";
import { getFieldValue } from "./utils/getFieldValue";

export function ResultField({
  name,
  value,
  topLine,
}: {
  name: string;
  value: number;
  topLine?: boolean;
}) {
  const { armenSalary, nastyaSalary } = useContext(ConfigContext);

  const values = getFieldValue(armenSalary.value, nastyaSalary.value, value);

  const nameStyle = topLine ? `${style.name} ${style.topLine}` : style.name;
  const valueStyle = topLine ? `${style.value} ${style.topLine}` : style.value;

  return values ? (
    <>
      <p className={nameStyle}>
        {topLine ? "" : "On "}
        {name}:
      </p>
      <p className={valueStyle}>[A] {values.firstValue}</p>
      <p className={valueStyle}>[N] {values.secondValue}</p>
    </>
  ) : (
    <></>
  );
}
