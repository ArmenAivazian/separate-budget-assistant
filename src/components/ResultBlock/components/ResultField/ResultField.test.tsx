import { createContext } from "react";
import renderer from "react-test-renderer";
import { ConfigContextWrapper } from "../../../../context/ConfigContextWrapper";
import { ResultField } from "./ResultField";
import { getFieldValue } from "./utils/getFieldValue";

const ConfigContext = createContext({
  armenSalary: { value: "1000" },
  nastyaSalary: { value: "250" },
});

describe("BlockResult -> Field", () => {
  // describe("snapshots", () => {
  //   it("default", () => {
  //     const component = renderer.create(
  //       <ConfigContext.Provider
  //         value={{
  //           armenSalary: { value: "1000" },
  //           nastyaSalary: { value: "250" },
  //         }}
  //       >
  //         <Field name="name" value={2} />
  //       </ConfigContext.Provider>
  //     );
  //     expect(component).toMatchSnapshot();
  //   });
  //   it("withTopLine", () => {
  //     const component = renderer.create(
  //       <ConfigContextWrapper>
  //         <Field name="name" value={2} topLine />
  //       </ConfigContextWrapper>
  //     );
  //     expect(component).toMatchSnapshot();
  //   });
  // });
  // describe("getFieldValue", () => {
  //   it("correct", () => {
  //     const result = getFieldValue("1000", "250", 1000);
  //     expect(result).toEqual({
  //       firstValue: 1000 * 0.75,
  //       secondValue: 1000 * 0.25,
  //     });
  //   });
  // });
});
