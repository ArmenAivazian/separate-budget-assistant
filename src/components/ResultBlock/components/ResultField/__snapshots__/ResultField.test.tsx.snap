// Vitest Snapshot v1

exports[`BlockResult -> Field > snapshots > default 1`] = `null`;

exports[`BlockResult -> Field > snapshots > withTopLine 1`] = `null`;
